assembly/src/parsers/ast/serialization.rs

```diff
@@ -0,0 +1,1165 @@
+    Dup11 = 118,
+    Dup12 = 119,
+    Dup13 = 120,
+    Dup14 = 121,
+    Dup15 = 122,
+    DupW0 = 123,
+    DupW1 = 124,
+    DupW2 = 125,
+    DupW3 = 126,
+    Swap = 127,
+    Swap2 = 128,
+    Swap3 = 129,
+    Swap4 = 130,
+    Swap5 = 131,
+    Swap6 = 132,
+    Swap7 = 133,
+    Swap8 = 134,
+    Swap9 = 135,
+    Swap10 = 136,
+    Swap11 = 137,
+    Swap12 = 138,
+    Swap13 = 139,
+    Swap14 = 140,
+    Swap15 = 141,
+    SwapW = 142,
+    SwapW2 = 143,
+    SwapW3 = 144,
+    SwapDW = 145,
+    MovUp2 = 146,
+    MovUp3 = 147,
+    MovUp4 = 148,
+    MovUp5 = 149,
+    MovUp6 = 150,
+    MovUp7 = 151,
+    MovUp8 = 152,
+    MovUp9 = 153,
+    MovUp10 = 154,
+    MovUp11 = 155,
+    MovUp12 = 156,
+    MovUp13 = 157,
+    MovUp14 = 158,
+    MovUp15 = 159,
+    MovUpW2 = 160,
+    MovUpW3 = 161,
+    MovDn2 = 162,
+    MovDn3 = 163,
+    MovDn4 = 164,
+    MovDn5 = 165,
+    MovDn6 = 166,
+    MovDn7 = 167,
+    MovDn8 = 168,
+    MovDn9 = 169,
+    MovDn10 = 170,
+    MovDn11 = 171,
+    MovDn12 = 172,
+    MovDn13 = 173,
+    MovDn14 = 174,
+    MovDn15 = 175,
+    MovDnW2 = 176,
+    MovDnW3 = 177,
+    CSwap = 178,
+    CSwapW = 179,
+    CDrop = 180,
+    CDropW = 181,
+
+    // ----- input / output operations --------------------------------------------------------
+    Adv = 182,
+    Locaddr = 183,
+    Sdepth = 184,
+    MemLoad = 185,
+    MemLoadImm = 186,
+    MemLoadW = 187,
+    MemLoadWImm = 188,
+    LocLoad = 189,
+    LocLoadW = 190,
+    MemStore = 191,
+    MemStoreImm = 192,
+    LocStore = 193,
+    MemStoreW = 194,
+    MemStoreWImm = 195,
+    LocStoreW = 196,
+    LoadWAdv = 197,
+
+    PushConstants = 198,
+    AdvU64Div = 199,
+    AdvPush = 200,
+    AdvLoadW = 201,
+
+    // ----- cryptographic operations ---------------------------------------------------------
+    RPHash = 202,
+    RPPerm = 203,
+    MTreeGet = 204,
+    MTreeSet = 205,
+    MTreeCWM = 206,
+
+    // ----- exec / call ----------------------------------------------------------------------
+    ExecLocal = 207,
+    ExecImported = 208,
+    CallLocal = 209,
+    CallImported = 210,
+}
+
+pub struct ByteWriter {
+    vec_bytes: Vec<u8>,
+}
+
+impl ByteWriter {
+    pub fn new() -> Self {
+        let vec_bytes = Vec::new();
+        Self { vec_bytes }
+    }
+
+    pub fn write_bool(&mut self, val: bool) {
+        self.write_u8(val as u8);
+    }
+
+    pub fn write_u8(&mut self, val: u8) {
+        self.vec_bytes.push(val);
+    }
+
+    pub fn write_u16(&mut self, val: u16) {
+        self.vec_bytes.append(&mut val.to_le_bytes().to_vec());
+    }
+
+    pub fn write_u32(&mut self, val: u32) {
+        self.vec_bytes.append(&mut val.to_le_bytes().to_vec());
+    }
+
+    pub fn write_u64(&mut self, val: u64) {
+        self.vec_bytes.append(&mut val.to_le_bytes().to_vec());
+    }
+
+    pub fn write_usize(&mut self, val: usize) {
+        self.vec_bytes
+            .append(&mut (val as u64).to_le_bytes().to_vec());
+    }
```

Comment:
I don't think we should write `usize` types. For strings, we should probably assume no more than 100 chars (and return an error if the string is longer), and for number of `repeat` iterations, a `u16` should probably be fine.